[libcxx] Improve libcxx tests when using Optimizations.

Stylie777 opened the pull request:
Within the libcxx tests, some tests compare a pointer to a Global Variable to one that is stored locally. At high levels of optmization the pointers are the same, but the assert is failing as the compiler cannot properly compare them. 

This changes how these variables are defined in specific tests to ensure the asserts pass. This is done by 
1. Updating a variant of `DoNotOptimize` used to accept L value references. This is a different variant that is currently used to ensure that the compiler cannot follow the data-flow through it by using both the input and output from the inline asm
2. Add in missing `DoNotOptimize` calls to tests. When using -O2 optimisations, these are failing because the pointer comparison fails when both values are the same. Adding the function call fixes this.

No change to the tests functionality is included.

Stylie777 committed:
[libcxx] Update `DoNotOptimize` to utilize L Value Reference Variant.

As part of the libcxx tests, there is a function provided called `DoNotOptimize`.
The tests currently utilise a version which only uses the value as an input,
leading to the compiler not removing information relating to the data-flow for
the value. This leads to issues when comparing the pointer values, where one is a
global pointer and the other a pointer which is local to the function being
executed.

Another version of `DoNotOptimize` is available which allows for the value
to be an input and output of the inline assembly, which allows for the
pointers to be successfully compared to one another. This function however
only accepts an L Value reference, where the values are initialized as R
value reference, so use the variant of `DoNotOptmize` described in the first
paragraph.

To enable the use of the second variant in the tests, the type of reference
has been updated within the test to be an L Value reference, to correctly
utlize the correct variant of `DoNotOptimize` that allows for successful
comparison of the pointers.

ldionne commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:
```diff
@@ -51,7 +51,8 @@ int main(int, char**) {
     {
         new_called = delete_called = 0;
-        OverAligned* x = DoNotOptimize(new OverAligned[3]);
+        OverAligned* dummy_data_block = new (std::nothrow) OverAligned;
```
This change uses the nothrow variant of `operator new`, but that is not what we are trying to test here.

ldionne commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp:
```diff
@@ -27,16 +27,31 @@
 #include "test_macros.h"
 #include "../types.h"
 
+class Data{
```
I don't understand why this change is necessary.

ldionne reviewed (changes requested):
Sorry if I am missing something obvious, but I basically don't understand this change. You mention that "the compiler cannot properly compare the values", what do you mean by that?

What issue are you encountering and with what configuration? Can you share a reproducer showcasing the issue?

Stylie777 committed:
[libcxx] Add `DoNotOptimize` function call to libcxx tests where appropriate

It has been seen, at high levels of optimizations, issues with comparing the
pointer values in some libcxx tests attaining to the assert comparing the pointer
of a global variable to that of one created locally within the test. These tests
follow the same logic as others that utilize the `DoNotOptimize` function, however
these tests do not.

To ensure the correct behaviour at high levels of optimization, these tests now utilize
the function when building the tests. This allows for the correct behaviour to be
observed when running the tests.

Stylie777 commented:
> Sorry if I am missing something obvious, but I basically don't understand this change. You mention that "the compiler cannot properly compare the values", what do you mean by that?

@ldionne  Apologies, not sure I explained it very well there. 

The issue that has been observed is that with certain configurations when using C++17, when comparing the pointers of `x` and `DummyData`, they result was coming out as false when the pointers are identical (see below). This was occurring on 6 tests, 3 of which can be remedied by utilising a different `DoNotOptimize` function that uses the input and output from the inline asm. The other three I used the class to define the DummyData, as then this removes the issues that can arise from comparing a global pointer to that of a local one. 

Example:
```
0000000000060940 <- `x`
0000000000060940 <- `DummyData`
0 <- (x == DummyData)
```

This is running on AArch64 V8 with the following compile options: `--target=aarch64-none-elf -march=armv8 -std=c++17 -O2`

jthackray reviewed (approved):
LGTM, once you've clang-formatted the code.

Stylie777 commented:
Updated. Ran Clang-Format

ldionne commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:
```diff
@@ -51,7 +51,8 @@ int main(int, char**) {
     {
         new_called = delete_called = 0;
-        OverAligned* x = DoNotOptimize(new OverAligned[3]);
+        OverAligned* dummy_data_block = new OverAligned;
```
This is still incorrect. This test is trying to test `operator new[]`, but you are not calling it anymore. You are calling the non-array version.

ldionne commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp:
```diff
@@ -27,16 +27,28 @@
+  std::size_t getDummyDataSize() { return sizeof(this->dummy_data_); }
+
+private:
+  alignas(OverAligned) char dummy_data_[alignof(OverAligned) * 3];
```
I don't understand why indirecting through a class solves anything here. After all you're still returning the address of a global variable.

Stylie777 replied:
I have removed the classes and reverted to the initial behaviour. Please see the initial comment as the changes made are different to those previously reviewed. Thanks

Stylie777 committed:
[NFC] Formatting Changes

ldionne commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_align_nothrow.replace.pass.cpp:
```diff
@@ -47,7 +47,7 @@ int main(int, char**) {
     {
         new_nothrow_called = delete_called = 0;
-        OverAligned* x = new (std::nothrow) OverAligned;
+        OverAligned* x                     = DoNotOptimize(new (std::nothrow) OverAligned);
```
Ah! I understand this part of the change -- that looks like an oversight to me. However, I don't understand why you need to add a rvalue-reference overload to the function below, it seems to me that the `const&` overload should handle this correctly, no?

Stylie777 replied:
The change to `&&value` was not the correct change, so have reverted this back to `&value` for `DoNotOptimize`.

In these test cases, we have found that the `const&` overload was not allowing for comparison of the pointers correctly, even when the values were the same. See explanation from @ostannard below:

"think that's because DoNotOptimize only uses the value as an input. This prevents the compiler from removing and data-flow leading up to it, but it can still "see" that one pointer came from new, and the other came from the address of a global, so they "can't" alias
This works for me, using both input and output from the inline asm, so the compiler can't follow the data-flow through it:"
```
template <class Tp>
inline Tp DoNotOptimize(Tp value) {
    asm volatile("" : "+r,m"(value) : : "memory");
    return value;
}
```

I have now scaled back the changes to only apply to the tests rather than change the function itself.

Stylie777 commented on libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:
```diff
@@ -51,7 +51,8 @@ int main(int, char**) {
     {
         new_called = delete_called = 0;
-        OverAligned* x = DoNotOptimize(new OverAligned[3]);
+        OverAligned* dummy_data_block = new OverAligned[3];
```
Yes we do - hence the change in these tests to a different variant of the `DoNotOptimize` function.

Stylie777 replied:
Updated

Stylie777 commented:
Thanks for merging @ldionne and thank you for the review. Much appreciated!
